pkg/cluster/cluster.go

```diff
@@ -0,0 +1,277 @@
+func (p *Peer) Leave(timeout time.Duration) error {
+	close(p.stopc)
+	return p.mlist.Leave(timeout)
+}
+
+// Name returns the unique ID of this peer in the cluster.
+func (p *Peer) Name() string {
+	return p.mlist.LocalNode().Name
+}
+
+// Peers returns a sorted address list of peers of the given type.
+func (p *Peer) Peers(t PeerType) (ps []string) {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	for _, o := range p.mlist.Members() {
+		os, ok := p.data[o.Name]
+		if !ok || os.Type != t {
+			continue
+		}
+		ps = append(ps, o.Address())
+	}
+	sort.Strings(ps)
+	return ps
+}
+
+// ClusterSize returns the current number of alive members in the cluster.
+func (p *Peer) ClusterSize() int {
+	return p.mlist.NumMembers()
+}
+
+// Info returns a JSON-serializable dump of cluster state.
+// Useful for debug.
+func (p *Peer) Info() map[string]interface{} {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	d := map[string]peerState{}
+	for k, v := range p.data {
+		d[k] = v
+	}
+
+	return map[string]interface{}{
+		"self":    p.mlist.LocalNode(),
+		"members": p.mlist.Members(),
+		"n":       p.mlist.NumMembers(),
+		"state":   d,
+	}
+}
+
+type peerState struct {
+	Type PeerType
+}
+
+// delegate implements memberlist.Delegate and memberlist.EventDelegate
+// and broadcasts its peer's state in the cluster.
+type delegate struct {
+	*Peer
+
+	logger log.Logger
+	bcast  *memberlist.TransmitLimitedQueue
+}
+
+func newDelegate(l log.Logger, p *Peer) *delegate {
+	bcast := &memberlist.TransmitLimitedQueue{
+		NumNodes:       p.ClusterSize,
+		RetransmitMult: 3,
+	}
+	return &delegate{
+		logger: l,
+		Peer:   p,
+		bcast:  bcast,
+	}
+}
+
+func (d *delegate) init(self string, numMembers func() int) {
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+
+	d.bcast = &memberlist.TransmitLimitedQueue{
+		NumNodes:       numMembers,
+		RetransmitMult: 3,
+	}
+}
+
+// NodeMeta retrieves meta-data about the current node when broadcasting an alive message.
+func (d *delegate) NodeMeta(limit int) []byte {
+	return []byte{}
+}
+
+// NotifyMsg is the callback invoked when a user-level gossip message is received.
+func (d *delegate) NotifyMsg(b []byte) {
+	var data map[string]peerState
+	if err := json.Unmarshal(b, &data); err != nil {
+		level.Error(d.logger).Log("method", "NotifyMsg", "b", strings.TrimSpace(string(b)), "err", err)
+		return
+	}
+	d.mtx.Lock()
+	defer d.mtx.Unlock()
+	for k, v := range data {
+		// Removing data is handled by NotifyLeave
+		d.data[k] = v
+	}
+}
+
+// GetBroadcasts is called when user data messages can be broadcasted.
+func (d *delegate) GetBroadcasts(overhead, limit int) [][]byte {
+	d.mtx.RLock()
+	defer d.mtx.RUnlock()
+	return d.bcast.GetBroadcasts(overhead, limit)
+}
+
+// LocalState
+func (d *delegate) LocalState(join bool) []byte {
+	d.mtx.RLock()
+	defer d.mtx.RUnlock()
+
+	b, err := json.Marshal(&d.data)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+func (d *delegate) MergeRemoteState(buf []byte, join bool) {
```

Comment:
Is actually join used?

Reply:
Ok it actually needs to be implement some interface.. please do `_ bool` if you don't use join